File: grokcpp/visual_struct/grok_structs/grok_heap.cpp
//
//  grok_heap.cpp
//  overload.cpp
//
//  Created by 胡泽弘 on 2022/11/17.
//

#include "grok_base.hpp"
using namespace std;

template <class _Tp, class _Compare = less<_Tp>>
class grok_Heap{
public:
    grok_Heap():_n(0),_hrm(new grok_base::head()){
        delete _hrm.sendto((new grok_base::act(1,0))->push("Heap"));
    };
    grok_Heap(const vector<_Tp> & arr):_n(arr.size()),_hrm(new grok_base::head()){
        delete _hrm.sendto((new grok_base::act(1,0))->push("Heap"));
        for(int i = 0;i < _n;i++)push(arr[i]);
    }
    void push(_Tp x){
        if(_n == _vec.size())_vec.push_back(0);
        insert(x, _n++);
    }
    _Tp pop(){
        swap(front(), _vec[--_n]);
        int i = 0;_Tp cur = _vec[0];
        for(int j = 1;j < _n;j<<=1, j++){
            if(j + 1 < _n && _Compare()(_vec[j], _vec[j+1]))j++;
            if(!_Compare()(cur, _vec[j]))break;
            _vec[i] = _vec[j];
            i = j;
        }
        _vec[i] = cur;
        delete _hrm.sendto(new grok_base::act(4,0));
        return _vec[_n];
    }
    _Tp & front(){
        if(empty())perror("Heap is empty");
        return _vec[0];
    }
    bool empty(){
        return _n == 0;
    }
    ~grok_Heap(){
        delete _hrm.sendto(new grok_base::act(-1,0));
    }
private:
    void insert(_Tp x,size_t i){
        while(i > 0 &&  _Compare()(_vec[(i-1)>>1],x)){
            _vec[i] = _vec[(i-1)>>1];
            (--i)>>=1;
        }
        _vec[i] = x;
        delete _hrm.sendto((new grok_base::act(3,1))->push(x));
    }
    vector<_Tp> _vec;
    size_t _n;
    grok_base::msg _hrm;
};

File: PyGrok/grok_heap.py
from PyGrokbase import *
from heapq import *


class heap:
    def __init__(self, arr=None):
        self._hrm = msg(head())
        self.heap = list()
        self._n = 0
        if arr is None:
            arr = []
        if len(arr) == 0:
            self._hrm.sendto(act(1, 0).push("heap"))
        else:
            self._hrm.sendto(act(1, 0).push("heap"))
            self._n = len(arr)
            for _i in range(self._n):
                self.push(arr[_i])

    def push(self, x):
        heappush(self.heap, x)
        self._n += 1
        self._hrm.sendto(act(3, 1).push(x))

    def pop(self):
        heappop(self.heap)
        self._n -= 1
        self._hrm.sendto(act(4, 0))

    def front(self):
        if self.empty():
            print("Heap is empty")
        return self.heap[0]

    def empty(self):
        return self._n == 0

    def __del__(self):
        self._hrm.sendto(act(-1, 0))

    def __str__(self):
        return str(self.heap)


if __name__ == '__main__':
    arr = [1, 4, 0, 10, 20, 99, 20, 7]
    h1 = heap(arr)
    print(h1)
